diff -urN --show-c-function alsa-driver-git20100411-hda/alsa-kernel/pci/hda/patch_conexant.c alsa-driver-git20100411-smartaudio/alsa-kernel/pci/hda/patch_conexant.c
--- alsa-driver-git20100411-hda/alsa-kernel/pci/hda/patch_conexant.c	2010-04-11 14:51:04.000000000 -0400
+++ alsa-driver-git20100411-smartaudio/alsa-kernel/pci/hda/patch_conexant.c	2010-04-11 15:01:00.000000000 -0400
@@ -557,6 +557,29 @@ static inline int conexant_init_jacks(st
 
 static void *conexant_hsfmodem_interface_funcs[];
 
+static void cxt_smartaudio_notify(unsigned int hp_present, int bios_profile)
+{
+	char *argv[4], *envp[3] = { "HOME=/", "PATH=/sbin:/bin:/usr/sbin:/usr/bin", 0 };
+
+	argv[0] = "/usr/bin/smartaudiocfg";
+	argv[1] = hp_present ? "--hp_present=1" : "--hp_present=0";
+	if(bios_profile >= 0) {
+		char bios_profile_buf[20];
+
+		snprintf(bios_profile_buf, sizeof(bios_profile_buf), "--bios_profile=%d", bios_profile);
+
+		argv[2] = bios_profile_buf;
+		argv[3] = 0;
+	} else
+		argv[2] = 0;
+
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,0)
+	call_usermodehelper(argv[0], argv, envp);
+#else
+	call_usermodehelper(argv[0], argv, envp, 1);
+#endif
+}
+
 static int conexant_init(struct hda_codec *codec)
 {
 	struct conexant_spec *spec = codec->spec;
@@ -812,6 +835,7 @@ static int conexant_build_controls(struc
 			return err;
 	}
 
+	cxt_smartaudio_notify(0, snd_hda_codec_read(codec, 0x01, 0, AC_VERB_GET_CONFIG_DEFAULT, 0) & 0xff);
 	return 0;
 }
 
@@ -1060,6 +1084,8 @@ static void cxt5045_hp_automute(struct h
 
 	spec->hp_present = snd_hda_jack_detect(codec, 0x11);
 
+	cxt_smartaudio_notify(spec->hp_present, -1);
+
 	bits = (spec->hp_present || !spec->cur_eapd) ? HDA_AMP_MUTE : 0; 
 	snd_hda_codec_amp_stereo(codec, 0x10, HDA_OUTPUT, 0,
 				 HDA_AMP_MUTE, bits);
@@ -1552,6 +1578,8 @@ static void cxt5047_hp_automute(struct h
 
 	spec->hp_present = snd_hda_jack_detect(codec, 0x13);
 
+	cxt_smartaudio_notify(spec->hp_present, -1);
+
 	bits = (spec->hp_present || !spec->cur_eapd) ? HDA_AMP_MUTE : 0;
 	/* See the note in cxt5047_hp_master_sw_put */
 	snd_hda_codec_amp_stereo(codec, 0x1d, HDA_OUTPUT, 0x01,
@@ -1984,6 +2012,9 @@ static void cxt5051_hp_automute(struct h
 	struct conexant_spec *spec = codec->spec;
 
 	spec->hp_present = snd_hda_jack_detect(codec, 0x16);
+
+	cxt_smartaudio_notify(spec->hp_present, -1);
+
 	cxt5051_update_speaker(codec);
 }
 
